File: sdk/beacon_sdk/context.py
from __future__ import annotations

from contextvars import ContextVar, Token
from dataclasses import dataclass

from beacon_sdk.models import Span


@dataclass
class TraceContext:
    """Holds the current trace_id and active span_id."""

    trace_id: str
    span_id: str | None = None


_trace_context: ContextVar[TraceContext | None] = ContextVar(
    "beacon_trace_context", default=None
)

# Active span registry — needed for get_current_span()
_active_spans: dict[str, Span] = {}


def get_context() -> TraceContext | None:
    return _trace_context.get()


def set_context(ctx: TraceContext) -> Token[TraceContext | None]:
    return _trace_context.set(ctx)


def reset_context(token: Token[TraceContext | None]) -> None:
    _trace_context.reset(token)


def register_span(span: Span) -> None:
    _active_spans[span.span_id] = span


def unregister_span(span_id: str) -> None:
    _active_spans.pop(span_id, None)


def get_active_span(span_id: str) -> Span | None:
    return _active_spans.get(span_id)

File: sdk-js/src/context.ts
/**
 * Trace context propagation using AsyncLocalStorage.
 *
 * Mirrors Python SDK's beacon_sdk/context.py but uses Node.js
 * AsyncLocalStorage instead of ContextVar for automatic async
 * context propagation.
 */

import { AsyncLocalStorage } from "node:async_hooks";
import type { Span } from "./models.js";

export interface TraceContext {
  traceId: string;
  spanId: string;
}

const asyncLocalStorage = new AsyncLocalStorage<TraceContext>();
const activeSpans = new Map<string, Span>();

export function getContext(): TraceContext | undefined {
  return asyncLocalStorage.getStore();
}

export function runWithContext<T>(ctx: TraceContext, fn: () => T): T {
  return asyncLocalStorage.run(ctx, fn);
}

export function registerSpan(span: Span): void {
  activeSpans.set(span.span_id, span);
}

export function unregisterSpan(spanId: string): void {
  activeSpans.delete(spanId);
}

export function getActiveSpan(spanId: string): Span | undefined {
  return activeSpans.get(spanId);
}

export function getCurrentSpan(): Span | undefined {
  const ctx = getContext();
  if (!ctx) return undefined;
  return activeSpans.get(ctx.spanId);
}
